ports/esp32s2/common-hal/busio/UART.c

```diff
@@ -0,0 +1,345 @@
+    switch (parity) {
+        case BUSIO_UART_PARITY_NONE:
+            parity_mode = UART_PARITY_DISABLE;
+            break;
+        case BUSIO_UART_PARITY_EVEN:
+            parity_mode = UART_PARITY_EVEN;
+            break;
+        case BUSIO_UART_PARITY_ODD:
+            parity_mode = UART_PARITY_ODD;
+            break;
+        default:
+            // Won't reach here because the input is an enum that is completely handled.
+            break;
+    }
+    uart_set_parity(self->uart_num, parity_mode);
+
+    // Stop is 1 or 2 always.
+    uart_stop_bits_t stop_bits= UART_STOP_BITS_1;
+    if (stop == 2) {
+        stop_bits = UART_STOP_BITS_2;
+    }
+    uart_set_stop_bits(self->uart_num, stop_bits);
+
+    self->tx_pin = NULL;
+    self->rx_pin = NULL;
+    self->rts_pin = NULL;
+    self->cts_pin = NULL;
+    int tx_num = -1;
+    int rx_num = -1;
+    int rts_num = -1;
+    int cts_num = -1;
+    if (have_tx) {
+        claim_pin(tx);
+        self->tx_pin = tx;
+        tx_num = tx->number;
+    }
+
+    if (have_rx) {
+        claim_pin(rx);
+        self->rx_pin = rx;
+        rx_num = rx->number;
+    }
+
+    if (have_rts) {
+        claim_pin(rts);
+        self->rts_pin = rts;
+        rts_num = rts->number;
+    }
+
+    if (have_cts) {
+        claim_pin(cts);
+        self->cts_pin = cts;
+        cts_num = cts->number;
+    }
+
+    if (have_rs485_dir) {
+        claim_pin(rs485_dir);
+        // RTS is used for RS485 direction.
+        self->rts_pin = rs485_dir;
+        rts_num = rs485_dir->number;
+    }
+    if (uart_set_pin(self->uart_num, tx_num, rx_num, rts_num, cts_num) != ESP_OK) {
+        mp_raise_ValueError(translate("Invalid pins"));
+    }
+}
+
+bool common_hal_busio_uart_deinited(busio_uart_obj_t *self) {
+    return self->rx_pin == NULL && self->tx_pin == NULL;
+}
+
+void common_hal_busio_uart_deinit(busio_uart_obj_t *self) {
+    if (common_hal_busio_uart_deinited(self)) {
+        return;
+    }
+    uart_driver_delete(self->uart_num);
+
+    common_hal_reset_pin(self->rx_pin);
+    common_hal_reset_pin(self->tx_pin);
+    common_hal_reset_pin(self->rts_pin);
+    common_hal_reset_pin(self->cts_pin);
+    self->rx_pin = NULL;
+    self->tx_pin = NULL;
+    self->cts_pin = NULL;
+    self->rts_pin = NULL;
+}
+
+// Read characters.
+size_t common_hal_busio_uart_read(busio_uart_obj_t *self, uint8_t *data, size_t len, int *errcode) {
+    if (self->rx_pin == NULL) {
+        mp_raise_ValueError(translate("No RX pin"));
+    }
+    if (len == 0) {
+        // Nothing to read.
+        return 0;
+    }
+
+    size_t total_read = 0;
+    uint64_t start_ticks = supervisor_ticks_ms64();
+
+    // Busy-wait until timeout or until we've read enough chars.
+    while (supervisor_ticks_ms64() - start_ticks <= self->timeout_ms) {
+        // Read as many chars as we can right now, up to len.
+        size_t num_read = uart_read_bytes(self->uart_num, data, len, 0);
```

Comment:
uart_read_bytes can return -1 for errors https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/uart.html#_CPPv415uart_read_bytes11uart_port_tP7uint8_t8uint32_t10TickType_t

if it does, the adjustments of total_read, etc., just below are incorrect.

It's not clear when/why this function _could_ error

Reply:
Ok, I exit the loop now if it is < 0.